python/cudf/udf_cpp/groupby/function.cu

```diff
@@ -0,0 +1,623 @@
+      T load    = data[threadIdx.x + item * blockDim.x];
+      local_min = min(local_min, load);
+    }
+  }
+
+  __syncthreads();
+
+  // Calculate local min for each group
+  atomicMin(smin, local_min);
+
+  __syncthreads();
+}
+
+// Use a C++ templated __device__ function to implement the body of the algorithm.
+template <typename T>
+__device__ void device_idxmax(T const* data,
+                              int const items_per_thread,
+                              int64_t const* index,
+                              size_type size,
+                              T init_val,
+                              T* smax,
+                              int64_t* sidx)
+{
+  // Calculate how many elements each thread is working on
+  T local_max       = init_val;
+  int64_t local_idx = -1;
+
+// Calculate local max for each thread
+#pragma unroll
+  for (size_type item = 0; item < items_per_thread; item++) {
+    if (threadIdx.x + (item * blockDim.x) < size) {
+      T load = data[threadIdx.x + item * blockDim.x];
+      if (load > local_max) {
+        local_max = load;
+        local_idx = index[threadIdx.x + item * blockDim.x];
+      }
+    }
+  }
+
+  __syncthreads();
+
+  // Calculate local max for each group
+  atomicMax(smax, local_max);
+
+  __syncthreads();
+
+  if (local_max == (*smax)) { atomicMin(sidx, local_idx); }
+
+  __syncthreads();
+}
+
+// Use a C++ templated __device__ function to implement the body of the algorithm.
+template <typename T>
+__device__ void device_idxmin(T const* data,
+                              int const items_per_thread,
+                              int64_t const* index,
+                              size_type size,
+                              T init_val,
+                              T* smin,
+                              int64_t* sidx)
+{
+  T local_min       = init_val;
+  int64_t local_idx = -1;
+
+// Calculate local max for each thread
+#pragma unroll
+  for (size_type item = 0; item < items_per_thread; item++) {
+    if (threadIdx.x + (item * blockDim.x) < size) {
+      T load = data[threadIdx.x + item * blockDim.x];
+      if (load < local_min) {
+        local_min = load;
+        local_idx = index[threadIdx.x + item * blockDim.x];
+      }
+    }
+  }
+
+  __syncthreads();
+
+  // Calculate local max for each group
+  atomicMin(smin, local_min);
+
+  __syncthreads();
+
+  if (local_min == (*smin)) { atomicMin(sidx, local_idx); }
+
+  __syncthreads();
+}
+
+extern "C" __device__ int BlockSum_int64(int64_t* numba_return_value,
+                                         int64_t const* data,
+                                         int64_t size)
+{
+  // Calculate how many elements each thread is working on
+  auto const items_per_thread = (size + blockDim.x - 1) / blockDim.x;
+
+  __shared__ int64_t sum;
+  if (threadIdx.x == 0) { sum = 0; }
+
+  __syncthreads();
+
+  device_sum<int64_t>(data, items_per_thread, size, &sum);
+
+  *numba_return_value = sum;
```

Comment:
Although it is a benign race, I would be happier if this were guarded by `if(threadIdx.x == 0) ...` I think, WDYT?